### PR TITLE
KAFKA-10048: Possible data gap for a consumer after a failover when u…

### PR DESCRIPTION
**Pull request by asdaraujo**
Changed the MM2 checkpoint mirror task to ensure it replicates consumer offsets even when they are equal to zero to avoid issues with consumers after failovers.

Modified the test case to cover the possible scenario of consumer gap, as described on KAFKA-10048. 

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

### CODE REVIEW

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java [hunk 1]
```diff
@@ -316,6 +354,23 @@ public void testReplication() throws InterruptedException {
             backup.kafka().consume(NUM_RECORDS_PRODUCED, 2 * RECORD_TRANSFER_DURATION_MS, "primary.test-topic-2").count());
     }
 
+    private Map<TopicPartition, List<ConsumerRecord<byte[], byte[]>>> consumeAllMessages(Consumer<byte[], byte[]> consumer) {
```
Comment (ryannedolan) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: Can we add a deadline to this function? Something like https://github.com/apache/kafka/blob/1c4eb1a5757df611735cfac9b709e0d80d0da4b3/connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/MirrorClient.java#L164

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java [hunk 2]
```diff
@@ -272,28 +303,35 @@ public void testReplication() throws InterruptedException {
...
+                .collect(Collectors.toList());
+        consumer2.assign(primaryPartitions);
         primaryOffsets.forEach(consumer2::seek);
-        consumer2.poll(Duration.ofMillis(500));
 
         assertTrue("Consumer failedback to zero upstream offset.", consumer2.position(new TopicPartition("test-topic-1", 0)) > 0);
         assertTrue("Consumer failedback to zero downstream offset.", consumer2.position(new TopicPartition("backup.test-topic-1", 0)) > 0);
         assertTrue("Consumer failedback beyond expected upstream offset.", consumer2.position(
-            new TopicPartition("test-topic-1", 0)) <= NUM_RECORDS_PRODUCED);
+            new TopicPartition("test-topic-1", 0)) <= Math.ceil((float) NUM_RECORDS_PRODUCED / (NUM_PARTITIONS - 1)) + Math.ceil((float) NUM_RECORDS_PRODUCED / NUM_PARTITIONS));
```
Comment (ryannedolan) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: I don't see why we need the cast, ceil, and coercion here? Integer math?

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java [hunk 3]
```diff
@@ -244,26 +251,50 @@ public void testReplication() throws InterruptedException {
...
+        assertEquals("Consumer failedover to non-zero offset on last partition.", 0,
+            consumer1.position(new TopicPartition("primary.test-topic-1", NUM_PARTITIONS - 1)));
         assertTrue("Checkpoints were not emitted upstream to primary cluster.", primary.kafka().consume(1,
             CHECKPOINT_DURATION_MS, "backup.checkpoints.internal").count() > 0);
 
+        Map<TopicPartition, List<ConsumerRecord<byte[], byte[]>>> messages1 = consumeAllMessages(consumer1);
+        System.out.println(messages1);
+        for (TopicPartition tp : backupPartitions) {
+            assertNotNull("No data consumed from partition " + tp + ".", messages1.get(tp));
+            int expectedMessageCount = tp.toString().equals("test-topic-1-0") ? 22 : 10;
```
Comment (ryannedolan) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: There is a lot of magic here -- why the special case and magic numbers? Can we restructure this to test the two cases separately maybe?

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java [hunk 4]
```diff
@@ -244,26 +251,50 @@ public void testReplication() throws InterruptedException {
...
         assertTrue("Consumer failedover beyond expected offset.", consumer1.position(
-            new TopicPartition("primary.test-topic-1", 0)) <= NUM_RECORDS_PRODUCED);
+            new TopicPartition("primary.test-topic-1", 0)) <= Math.ceil((float) NUM_RECORDS_PRODUCED / (NUM_PARTITIONS - 1)));
+        assertEquals("Consumer failedover to non-zero offset on last partition.", 0,
+            consumer1.position(new TopicPartition("primary.test-topic-1", NUM_PARTITIONS - 1)));
         assertTrue("Checkpoints were not emitted upstream to primary cluster.", primary.kafka().consume(1,
             CHECKPOINT_DURATION_MS, "backup.checkpoints.internal").count() > 0);
 
+        Map<TopicPartition, List<ConsumerRecord<byte[], byte[]>>> messages1 = consumeAllMessages(consumer1);
+        System.out.println(messages1);
```
Comment (ryannedolan) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: errant println

**Review by ryannedolan — CHANGES_REQUESTED** [review]
Thanks for discovering, verifying, and fixing this issue! Change makes sense to me.

Requested change: add deadline to consumeAllMessages() -- otherwise lgtm.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java (hunk 1 above)
Comment (asdaraujo) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: Good point. Will add that.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java (hunk 2 above)
Comment (asdaraujo) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: Yep. I need to cast to float to ensure that the result of the division is a float, with a non-zero decimal value, so that ceil will round it up to the nearest integer.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java (hunk 4 above)
Comment (asdaraujo) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: Oops... it escaped my final review. Removing it.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java (hunk 3 above)
Comment (asdaraujo) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: You're right. This looks magic indeed! I had to think about it again to understand my own reasoning :)
I added comments to explain what's going on here and why we expect 22 for one partition and 10 for the others.
I left the tests as they were. I think the comment makes the test logic clear.

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointTask.java [hunk 5]
```diff
@@ -132,7 +132,7 @@ public String version() {
             return listConsumerGroupOffsets(group).entrySet().stream()
                 .filter(x -> shouldCheckpointTopic(x.getKey().topic()))
                 .map(x -> checkpoint(group, x.getKey(), x.getValue()))
-                .filter(x -> x.downstreamOffset() > 0)  // ignore offsets we cannot translate accurately
+                .filter(x -> x.downstreamOffset() >= 0)  // ignore offsets we cannot translate accurately
```
Comment (heritamas) [inline connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointTask.java]: Can downstream offset be negative?

Comment (ryannedolan) [reply connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointTask.java]: nope, that's not possible

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java (hunk 1 above)
Comment (asdaraujo) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: This has been added.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java (hunk 4 above)
Comment (asdaraujo) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: Done

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointTask.java (hunk 5 above)
Comment (heritamas) [reply connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointTask.java]: Then, this filter matches everything, doesn't it?

Comment (asdaraujo) [reply connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointTask.java]: @ryannedolan @heritamas If you guys think this is safe enough I can remove that filter. But it doesn't hurt to leave it there just in case a rogue negative offset comes through for whatever reason/bug... Please let me know what you think.

Comment (heritamas) [reply connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorCheckpointTask.java]: Yes, does not hurt to leave it. Just for sure.

**Comment by viktorsomogyi** [comment]
@ijuma would you please review this or suggest who is the right committer to help with this?

**Comment by ijuma** [comment]
@mimaison, would you be able to review this?

**Comment by mimaison** [comment]
Sure I'll take a look this week

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java [hunk 6]
```diff
@@ -367,14 +406,37 @@ public void testOneWayReplicationWithAutorOffsetSync1() throws InterruptedExcept
...
 
         records = consumer.poll(Duration.ofMillis(500));
         // similar reasoning as above, no more records to consume by the same consumer group at backup cluster
         assertEquals("consumer record size is not zero", 0, records.count());
         consumer.close();
+    }
+
+    private void produceMessages(EmbeddedConnectCluster cluster, String topicName, int partitions, String msgPrefix) {
+        for (int i = 0; i < NUM_RECORDS_PRODUCED; i++) {
+            // produce to all partitions but the last one
```
Comment (mimaison) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: This comment needs updating

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java [hunk 7]
```diff
@@ -128,10 +136,23 @@ public void setup() throws InterruptedException {
         backup.kafka().createTopic("primary.test-topic-1", 1);
         backup.kafka().createTopic("heartbeats", 1);
 
-        for (int i = 0; i < NUM_RECORDS_PRODUCED; i++) {
-            primary.kafka().produce("test-topic-1", i % NUM_PARTITIONS, "key", "message-1-" + i);
-            backup.kafka().produce("test-topic-1", i % NUM_PARTITIONS, "key", "message-2-" + i);
-        }
+        // produce to all partitions but the last one
```
Comment (mimaison) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: Would it be better using a separate topic in order to keep a partition without any records? By changing this topic it affects existing checks in all tests

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java [hunk 8]
```diff
@@ -128,10 +136,23 @@ public void setup() throws InterruptedException {
...
+        produceMessages(primary, "test-topic-1", NUM_PARTITIONS - 1, "message-1-");
+        produceMessages(backup, "test-topic-1", NUM_PARTITIONS - 1, "message-2-");
+
+        consumerProps = new HashMap<String, Object>() {{
```
Comment (mimaison) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: As this does not change, I wonder if we could direct initialize `consumerProps` when it's declared

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java [hunk 9]
```diff
@@ -190,24 +211,19 @@ public void close() {
...
-        Consumer<byte[], byte[]> consumer1 = primary.kafka().createConsumerAndSubscribeTo(Collections.singletonMap(
-            "group.id", "consumer-group-1"), "test-topic-1", "backup.test-topic-1");
+        Consumer<byte[], byte[]> consumer1 = primary.kafka().createConsumerAndSubscribeTo(consumerProps, "test-topic-1", "backup.test-topic-1");
         consumer1.poll(Duration.ofMillis(500));
         consumer1.commitSync();
         consumer1.close();
 
-        Consumer<byte[], byte[]> consumer2 = backup.kafka().createConsumerAndSubscribeTo(Collections.singletonMap(
-            "group.id", "consumer-group-1"), "test-topic-1", "primary.test-topic-1");
+        Consumer<byte[], byte[]> consumer2 = backup.kafka().createConsumerAndSubscribeTo(consumerProps, "test-topic-1", "primary.test-topic-1");
```
Comment (mimaison) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: Do we still need these 2 blocks? In `setup()` we already consumed all messages

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java (hunk 3 above)
Comment (mimaison) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: I'm actually surprized we only see positions `22` and `10`. Why do we only get `test-topic-1-0` here and not the other 9 partitions?

**Review by mimaison — CHANGES_REQUESTED** [review]
Good catch, this is indeed a problem if some partitions don't have records before the failover.

I've left a few comments. Overall I find the test pretty hard to follow. I wonder if more comments or some refactoring would help.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java (hunk 6 above)
Comment (asdaraujo) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: Good catch. Updating.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java (hunk 7 above)
Comment (asdaraujo) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: Good point. I'll look into this. Besides not affecting the other tests, it should make it simpler to reason about

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java (hunk 6 above)
Comment (asdaraujo) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: Updated.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java (hunk 7 above)
Comment (asdaraujo) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: I've separated the test from the existing ones, also using a different topic. Some of the logic on those tests is complex and may be hard to follow so I thought it would be better to have the tests totally separate and simpler to interpret.

I think, and hope, it's easier to understand now than it was before.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java (hunk 8 above)
Comment (asdaraujo) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: I changed the way this is done, setting it at the test-level, with test-specific CGs.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java (hunk 9 above)
Comment (asdaraujo) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: Good catch. Also changed this and we're now consuming it only once.

**Comment by asdaraujo** [comment]
@mimaison Thanks for the feedback. I've refactored the tests. Could you please give it another review.

**Comment by mimaison** [comment]
Thanks @asdaraujo, I'll try to take another look later on this week.

Can you rebase on trunk? to pick up a change (https://github.com/apache/kafka/commit/241e1447fae8b85b3bb491f371357b8f92d2da72) that is now necessary to run the CI tests

**Comment by asdaraujo** [comment]
Thanks, @mimaison . I've rebased it.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java [hunk 10]
```diff
@@ -283,49 +296,140 @@ public void testReplication() throws InterruptedException {
...
+
+        // Consume all messages
+        consumeAllMessages(consumer, NUM_RECORDS_PER_PARTITION * (NUM_PARTITIONS - 1));
+        consumer.close();
+
+        // Consumer group offsets after consumption: topic's last partition doesn't yet has data, so
+        // the committed offset is 0. All other topic's partition should have offset equal to NUM_RECORDS_PER_PARTITION.
+        // backupTopicReplica still has a single empty partition, since MM2 is not yet started, and its record offset is 0.
+
+        mm2Config = new MirrorMakerConfig(mm2Props);
```
Comment (edoardocomar) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: this overwrite of mm2config should go in the setup method, IMHO

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java [hunk 11]
```diff
@@ -283,49 +296,140 @@ public void testReplication() throws InterruptedException {
...
+            backup.kafka().consume(NUM_RECORDS_PER_PARTITION, RECORD_TRANSFER_DURATION_MS, "test-topic-3").count());
+
+        assertEquals("New topic was not replicated to primary cluster.", NUM_RECORDS_PER_PARTITION,
+            primary.kafka().consume(NUM_RECORDS_PER_PARTITION, 2 * RECORD_TRANSFER_DURATION_MS, "backup.test-topic-3").count());
+        assertEquals("New topic was not replicated to backup cluster.", NUM_RECORDS_PER_PARTITION,
+            backup.kafka().consume(NUM_RECORDS_PER_PARTITION, 2 * RECORD_TRANSFER_DURATION_MS, "primary.test-topic-2").count());
+    }
+
+    @Test
+    public void testReplicationWithEmptyPartition() throws InterruptedException {
```
Comment (mimaison) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: This test is overly complicated. I think it could:
- Create a topic
- Produce messages to all partitions but one
- Consume all messages
- Start a single MirrorMaker2 instance primary->backup
- Use `RemoteClusterUtils.translateOffsets()` to retrieve offsets
- Assert offset for the last partition is 0

For example, something along these lines (this cuts a few corners so you'd need to improve it)

```suggestion
    @Test
    public void testReplicationWithEmptyPartition() throws Exception {
        String consumerGroupName = "consumer-group-testReplicationWithEmptyPartition";
        Map<String, Object> consumerProps  = new HashMap<String, Object>() {{
            put("group.id", consumerGroupName);
            put("auto.offset.reset", "earliest");
        }};

        String topic = "test-topic-empty";
        primary.kafka().createTopic(topic, NUM_PARTITIONS);
        mm2Config = new MirrorMakerConfig(mm2Props);
        // produce to all test-topic-empty's partitions *but the last one*, on the primary cluster
        produceMessages(primary, topic, "message-1-", NUM_PARTITIONS - 1);

        // Consume, from the primary cluster, before starting the connectors so we don't need to wait for discovery
        Consumer<byte[], byte[]> consumer = primary.kafka().createConsumerAndSubscribeTo(consumerProps, topic);
        consumeAllMessages(consumer, NUM_RECORDS_PER_PARTITION * (NUM_PARTITIONS - 1));
        consumer.close();

        waitUntilMirrorMakerIsRunning(backup, mm2Config, "primary", "backup");

        Map<TopicPartition, OffsetAndMetadata> backupOffsets = RemoteClusterUtils.translateOffsets(
                mm2Config.clientConfig("backup").adminConfig(),
                "primary",
                consumerGroupName,
                Duration.ofMillis(CHECKPOINT_DURATION_MS));

        OffsetAndMetadata oam = backupOffsets.get(new TopicPartition("primary." + topic, NUM_PARTITIONS - 1));
        assertNotNull(oam);
        assertEquals(0, oam.offset());
    }
```

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java [hunk 12]
```diff
@@ -136,10 +141,19 @@ public void setup() throws InterruptedException {
...
-        for (int i = 0; i < NUM_RECORDS_PRODUCED; i++) {
-            primary.kafka().produce("test-topic-1", i % NUM_PARTITIONS, "key", "message-1-" + i);
-            backup.kafka().produce("test-topic-1", i % NUM_PARTITIONS, "key", "message-2-" + i);
-        }
+        // produce to all partitions of test-topic-1
+        produceMessages(primary, "test-topic-1", "message-1-");
+        produceMessages(backup, "test-topic-1", "message-2-");
+
+        // Generate some consumer activity on both clusters to ensure the checkpoint connector always starts promptly
+        Map<String, Object> dummyProps = new HashMap<String, Object>();
```
Comment (mimaison) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: We can use `Collections.singletonMap()` here

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java [hunk 13]
```diff
@@ -201,26 +213,24 @@ public void close() {
 
     @Test
     public void testReplication() throws InterruptedException {
+        String consumerGroupName = "consumer-group-testReplication";
+        Map<String, Object> consumerProps  = new HashMap<String, Object>() {{
+                put("group.id", consumerGroupName);
+                put("auto.offset.reset", "latest");
```
Comment (mimaison) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java:219]: `latest` is the default, why are we setting it?

**Review by mimaison — CHANGES_REQUESTED** [review]
Thanks for the update. 
Considering the code change (a single line), the test is really complicated. Other existing tests already cover MM2 use cases. The new test could only focus on the difference of behaviour caused by the fix.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java (hunk 10 above)
Comment (asdaraujo) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: Thanks, @edoardocomar . I've addressed this.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java (hunk 12 above)
Comment (asdaraujo) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: Fixed.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java (hunk 11 above)
Comment (asdaraujo) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: Thanks, @mimaison . I've changed the test to simplify it as per your suggestion.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java (hunk 13 above)
Comment (asdaraujo) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java:219]: Yep, unnecessary. I'll remove this.

Comment (asdaraujo) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java:219]: Oops, spoke too soon.
Even though `latest` is Kafka's default, `EmbeddedKafkaCluster::createConsumer` [defaults it to `earliest`](https://github.com/apache/kafka/blob/48b56e533b3ff22ae0e2cf7fcc649e7df19f2b06/connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java#L444)

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java [hunk 14]
```diff
@@ -381,45 +440,46 @@ public void testOneWayReplicationWithAutoOffsetSync() throws InterruptedExceptio
 
         waitUntilMirrorMakerIsRunning(backup, mm2Config, "primary", "backup");
 
-        // create a consumer at backup cluster with same consumer group Id to consume 1 topic
-        Consumer<byte[], byte[]> consumer = backup.kafka().createConsumerAndSubscribeTo(
-            Collections.singletonMap("group.id", "consumer-group-1"), "primary.test-topic-1");
+        // Map<TopicPartition, OffsetAndMetadata> offsets =
```
Comment (mimaison) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: Is this left over from debugging?

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java [hunk 15]
```diff
@@ -429,4 +489,69 @@ private void deleteAllTopics(EmbeddedKafkaCluster cluster) {
         } catch (Throwable e) {
         }
     }
+
+    private void produceMessages(EmbeddedConnectCluster cluster, String topicName, String msgPrefix) {
+        produceMessages(cluster, topicName, msgPrefix, null);
```
Comment (mimaison) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java]: I wonder if we should pass `NUM_PARTITIONS` instead of `null` for the last argument. Then `numPartitions` can be an `int` in the other `produceMessages()` method. WDYT?

**Review by mimaison — CHANGES_REQUESTED** [review]
Thanks for the updates. I think we're almost there! I've left a couple of comments

**Review by mimaison — APPROVED** [review]

**Comment by ning2008wisc** [comment]
Hi @mimaison I will rebase my pr https://github.com/apache/kafka/pull/9224 on top of this one over this weekend and hope you may have time soon to review it as your next one. Thanks
